### ISSUE TITLE
Discuss optional dependencies for notebooks

### ISSUE THREAD

**Issue by fcooper8472** [pints-team/pints#1108]
Opening a dedicated issue because of the unresolved discussion [here](https://github.com/pints-team/pints/issues/1074#issuecomment-609423080).

tl;dr: one of the example notebooks uses `autograd` which is not a pints dependency, so the notebook was failing.

The question was whether to support the notebooks in pints, or whether to move them to a different repository.  I will attempt to summarise the key arguments (@MichaelClerx @ben18785 please feel free to jump in here):

**For keeping it in pints:**
- Natural place for people to look for examples
- More people are likely to find it if it's with the other examples

**Against keeping it in pints:**
- It introduces a new dependency

The current state of play is that I wrapped the import in a `try` block:

```python
try:
    import autograd.numpy as np
    from autograd.scipy.integrate import odeint
    from autograd.builtins import tuple
    from autograd import grad
except ImportError:
    print("""This example requires autograd, which is not a pints dependency.
    If you see this warning, try `pip install autograd`""")
    exit(0)
```

## Suggested solution

I suggest adding new section of `optional` or `suggested` dependencies:
```
extras_require={
    'docs': [
        'sphinx>=1.5, !=1.7.3',     # For doc generation
    ],
    'dev': [
        'flake8>=3',            # For code style checking
        'jupyter',              # For documentation and testing
        'nbconvert',
        'traitlets',
    ],
    'suggested': [
        'autograd',
    ]
},
```

Then a "user" of pints will never get anything optional - but we could encase any `suggested` dependencies in the same kind of `try` block so that a user will get a human readable error if they try to run the example without the necessary libraries.

Presumably the error message could ask the user to
```
pip install .[suggested]
```

This is a mechanism that we've used before with Chaste, which defines suggestions such as `paraview` that are useful but not essential.  This seems like the best of both worlds as we can "show off" cool things but without users being forced to install unnecessary dependencies if they just want the core functionality.

**Comment by martinjrobins** [comment]
I would like to keep the examples using external libraries in the pints repo. I quite like @fcooper8472 's solution as it is, using the try block and telling a user what to install to get the notebook working. Seems straightforward enough. And the CI build that tests the notebook can just do a `pip install autograd` so that it is tested.

**Comment by MichaelClerx** [comment]
Thanks for the summary Fergus!

I'm not sure you've captured the "for" argument well though:

> - Natural place for people to look for examples
> - More people are likely to find it if it's with the other examples

We'd still have both of those advantages if we had the notebooks in an external repository that we linked to from the examples page (which is my preferred option).

I'd suggest the "for" parts instead is:

> - we don't have to create an extra repo for each example that uses software that PINTS doesn't depend on
> - we can test the notebooks along with the other notebooks

I don't think these are good enough reasons for us to say "pints depends on X" when clearly it doesn't, or to say "PINTS suggests you also install Y" (I guess?). I'd be much happier to say "look we made this cool example, please see repo Z".

Also, where's the cut-off point? Are we going to say myokit and pybamm are "suggested" dependencies of pints now too, and drag those examples into the notebook folder? The whole point of the ForwardModel class is that you can use it with lots of other things, we can't add all of those to the "suggested" list of things you install.

Also, users encountering bugs will probably start installing all the suggested dependencies, just in case

**Comment by MichaelClerx** [comment]
Thanks for the summary Fergus! I'm happy for us to discuss this (although I think I'm being overruled on this one)

I'm not sure you've captured the "for" argument well though:

> - Natural place for people to look for examples
> - More people are likely to find it if it's with the other examples

We'd still have both of those advantages if we had the notebooks in an external repository that we linked to from the examples page (which is my preferred option).

I'd suggest the "for" part instead is:

> - we don't have to create an extra repo for each example that uses software that PINTS doesn't depend on
> - we can test the notebooks along with the other notebooks

and the "against" part is

> - It introduces a new dependency
> - It add things to the "dependencies" that PINTS does not depend on

I don't think these are good enough reasons for us to say "pints depends on X" when clearly it doesn't, or to say "PINTS suggests you also install Y" (I guess?). I'd be much happier to say "look we made this cool example, please see repo Z".

Also, where's the cut-off point? Are we going to say myokit and pybamm are "suggested" dependencies of pints now too, and drag those examples into the notebook folder? The whole point of the ForwardModel class is that you can use it with lots of other things, we can't add all of those to the "suggested" list of things you install.

Also, users encountering bugs will probably start installing all the suggested dependencies, just in case

**Comment by MichaelClerx** [comment]
Thanks for the summary Fergus! I'm happy for us to discuss this (although I think I'm being overruled on this one)

I'm not sure you've captured the "for" argument well though:

> - Natural place for people to look for examples
> - More people are likely to find it if it's with the other examples

We'd still have both of those advantages if we had the notebooks in an external repository that we linked to from the examples page (which is my preferred option).

I'd suggest the "for" part instead is:

> - we don't have to create an extra repo for each example that uses software that PINTS doesn't depend on
> - we can test the notebooks along with the other notebooks

and the "against" part is

> - It introduces a new dependency
> - It add things to the "dependencies" that PINTS does not depend on
> - It becomes the core PINTS team's responsibility to keep these examples working. As they require extra software, and are likely to do something showy, this is likely to cause some work for the maintainers

I don't think these are good enough reasons for us to say "pints depends on X" when clearly it doesn't, or to say "PINTS suggests you also install Y" (I guess?). I'd be much happier to say "look we made this cool example, please see repo Z".

Also, where's the cut-off point? Are we going to say myokit and pybamm are "suggested" dependencies of pints now too, and drag those examples into the notebook folder? The whole point of the ForwardModel class is that you can use it with lots of other things, we can't add all of those to the "suggested" list of things you install.

Also, users encountering bugs will probably start installing all the suggested dependencies, just in case

**Comment by MichaelClerx** [comment]
We can do these things with "interfaces" now #1270
